dali/pipeline/executor/executor.cc

```diff
@@ -644,7 +658,47 @@ bool Executor<WorkspacePolicy, QueuePolicy>::HasConditionals() const {
   return has_conditionals_;
 }
 
+template<typename WorkspacePolicy, typename QueuePolicy>
+void Executor<WorkspacePolicy, QueuePolicy>::InitCheckpointing() {
+  if (!checkpointing_)
+    return;
+
+  if (std::is_same_v<QueuePolicy, SeparateQueuePolicy>)
+    DALI_FAIL("Checkpointing is not supported with `separated` pipeline exection mode enabled. ")
 
+  checkpointing_epoch_size_ = -1;
+  for (const auto &node : graph_->GetOpNodes()) {
+    auto meta = node.op->GetReaderMeta();
+    if (meta.epoch_size_padded <= 0)
+      continue;
+
+    int local_epoch_size = (meta.epoch_size_padded + max_batch_size_ - 1) / max_batch_size_;
+    if (checkpointing_epoch_size_ == -1)
+      checkpointing_epoch_size_ = local_epoch_size;
+    else if (checkpointing_epoch_size_ != local_epoch_size)
+      DALI_FAIL(make_string(
+        "When the checkpointing is enabled, all readers must have the same epoch size. ",
+        "The readers and have different epoch sizes ",
```

Comment:
Something is missing in the message (`The readers and have`). In the first remark I meant for the message to mention names of both readers. You either need to include them in the message or reformulate the message a bit.